### ISSUE TITLE
In app fiat donation / subscription (apple pay)

### ISSUE THREAD

**Issue by alltheseas** [damus-io/damus#1422]
## user story
As a Damus user who appreciates the devs, I would like a way to donate to Damus in fiat, so that I can support the devs and help contribute to Damus maintenance & development costs.

### acceptance criteria
1. [x] Damus user can "subscribe" to Damus via fiat apple pay mechanism
2. [x] ~~Damus user can send a one time donation to Damus via fiat apple pay mechanism~~
     - Delegated to https://github.com/damus-io/damus/issues/1813
4. [x] Damus user can see their subscription status
5. [x] Damus user can unsubscribe from Damus purple
6. [x] Damus user receives star ⭐️ upon subscription
7. [x] There is annual subscription option
8. [x] ~~Show user that apple takes X%/Y$ of each in app apple pay donation~~ Delegated to https://github.com/damus-io/damus/issues/1814
9. [x] Connect in-app purchase with server registration
10. [x] There should be a "welcome" or "thank you" screen

~~12. [x] Optimize views to look good on all devices/screens
13. [x] Mention that first 1,000 annual subscribers will receive "Nº 1", "Nº 2" subscriber numbers
14. [x] Implement restoring the purchase, and syncing on different devices under the same App Store account
15. [x] Polish subscription mechanism so that it registers with the server immediately after subscribing (without having to restart the app)~~ Delegated to https://github.com/damus-io/damus/issues/1814

17. [X] ~~Implement secure receipt transmission with cryptographic signatures~~
    - Delegated to https://github.com/damus-io/damus/issues/1809
18. [X] ~~Implement secure receipt verification with the Apple server (Daniel: I might get blocked on this, requires some kind of private key)~~
    - Delegated to https://github.com/damus-io/damus/issues/1812
19. [X] Implement receipt verification mechanisms for both Sandbox and production environments
20. [X] ~~Test flow on a sandbox (and do all necessary configs to make it happen)~~
    - Delegated to https://github.com/damus-io/damus/issues/1810
21. [X] ~~Test and address any missing details on other flows such as cancellation and refunds~~
    - Delegated to https://github.com/damus-io/damus/issues/1811
22. [x] Wrap up work on ticket and make sure it is hidden behind a developer setting so that an average user does not have access to it while we work on it.

### how might we/questions
1. Should Damus send a receipt via DM to the subscriber?

### commentary
At first Damus purple does not need to add additional features (see purple #659 ). Additional features can be added over time.
Subscribe with bitcoin #1755

### related
Damus Purple https://github.com/damus-io/damus/issues/659

**Comment by alltheseas** [comment]
Recommended immediately post, or alongside nostrdb update @jb55

**Comment by alltheseas** [comment]
@jb55 advised preference is to stick with sats, and not take fiat.

**Comment by jb55** [comment]
On Tue, Nov 21, 2023 at 10:14:27AM -0800, alltheseas wrote:
>@jb55 advised preference is to stick with sats, and not take fiat.

this is not true

**Comment by jb55** [comment]
damus is not only for bitcoiners

**Comment by alltheseas** [comment]
Ok. I must have misunderstood.

**Comment by alltheseas** [comment]
prompt https://github.com/damus-io/damus/issues/1756

**Comment by alltheseas** [comment]
signal subscribe example

![image](https://github.com/damus-io/damus/assets/64376233/c3f7ee6a-db38-4069-93fd-74f86ef379b4)

signal one-time example
![image](https://github.com/damus-io/damus/assets/64376233/30b86385-760b-40a1-9277-f0815027491c)

**Comment by danieldaquino** [comment]
I started working on this feature! I will start with the overall screens and other items of the acceptance criteria, study Apple's in-app mechanism, and then integrate the payment mechanism at the end.

**Comment by jb55** [comment]
On Fri, Dec 01, 2023 at 03:40:33PM -0800, Daniel D’Aquino wrote:
>I started working on this feature! I will start with the overall
>screens and other items of the acceptance criteria, study Apple's
>in-app mechanism, and then integrate the payment mechanism at the end.

There's already a branch for this! Lemme revive it and see if there's
anything worth salvaging

**Comment by jb55** [comment]
On Fri, Dec 01, 2023 at 03:53:59PM -0800, William Casarin wrote:
>On Fri, Dec 01, 2023 at 03:40:33PM -0800, Daniel D’Aquino wrote:
>>I started working on this feature! I will start with the overall
>>screens and other items of the acceptance criteria, study Apple's
>>in-app mechanism, and then integrate the payment mechanism at the end.
>
>There's already a branch for this! Lemme revive it and see if there's
>anything worth salvaging

pushed to the purple branch

**Comment by danieldaquino** [comment]
Thank you @jb55, this was definitely very helpful!

I am building this feature on top of @jb55's previous work on this, and I have done the following changes so far:
1. Updated the copywriting to be more accurate to what we are planning to deliver with Purple on the short term
2. Updated the design to be increase consistency with Rob's designs
3. I am working on getting the star badge applied to users who subscribe to it, and although I have some code drafted up, it does not seem to be working yet, so still some debugging and fixing to get it done.

Here is how it currently looks like:

<img width="1552" alt="Screenshot 2023-12-02 at 02 30 22 copy" src="https://github.com/damus-io/damus/assets/24692108/96107fe2-69d5-4dad-b4d0-0ada353dbf11">

@jb55, @robagreda, @alltheseas, please feel free to provide any suggestions on the design if it can be improved!

I updated the acceptance criteria on the top to mark things I believe are ready. On top of that, a few other thoughts of things we might want to address (on this or on a separate ticket):
- [ ] Creating a website page for Damus Purple with an FAQ (where we can, for example, hint that they will be able to pay for Damus Purple in sats). Maybe we can also add a small roadmap of what they can expect in terms of upcoming premium features?
- [ ] Create a nice warm confirmation "thank you" view (maybe with some special effect), to add an extra moment of delight for people who just subscribed. We need to make sure we make them happy/surprised when they subscribe.

**Comment by alltheseas** [comment]
@danieldaquino mention that first 1,000 annual subscribers will receive a #1, #2 subscriber number

**Comment by danieldaquino** [comment]
@alltheseas, from today's standup, can I go ahead and make the following modifications to the acceptance criteria?
1. Remove _"Damus user receives star ⭐️ upon subscription"_
2. Add _"Connect in-app purchase with server registration"_

**Comment by alltheseas** [comment]
> can you please clarify what the expectation is regarding the ⭐️ on users' profiles?

I think @jb55 mentioned that the reason why we need server registration, is precisely for this reason - that anyone technical enough can add a ⭐, whereas a registration on a server specifically checks if the user has paid for Damus, and grants ⭐ as a reward.

**Comment by jb55** [comment]
On Mon, Dec 04, 2023 at 10:55:40AM -0800, alltheseas wrote:
>> can you please clarify what the expectation is regarding the ⭐️ on users' profiles?
>
>I think @jb55 mentioned that the reason why we need server
>registration, is precisely for this reason - that anyone technical
>enough can add a ⭐, whereas a registration on a server specifically
>checks if the user has paid for Damus, and grants ⭐ as a reward.

Correct. We are keeping the star but we're just making it more secure.
Damus needs to call out to the API with a list of npubs to check
registration statuses.

**Comment by jb55** [comment]
On Mon, Dec 04, 2023 at 11:07:33AM -0800, Daniel D’Aquino wrote:
>> I think @jb55 mentioned that the reason why we need server
>> registration, is precisely for this reason - that anyone technical
>> enough can add a ⭐, whereas a registration on a server specifically
>> checks if the user has paid for Damus, and grants ⭐ as a reward.
>
>That makes sense. @jb55, in order to check whether or not users have the star icon, do you prefer to:
>1. consult directly with the Damus server, or
>2. have the Damus server issue the "⭐" using [NIP-58](https://github.com/nostr-protocol/nips/blob/master/58.md) badges?
>
>I imagine (2) is better, but I would like to double-check if that's
>what you are looking for.

Badges aren't the right approach here, since they don't expire. I also
want our stars to be much smarter over time. It's not a static one time
thing. We can customize the look of profiles and stars based on
registration status and maybe even how long they have been a continuous
subscriber, like streak rewards.

**Comment by danieldaquino** [comment]
> Badges aren't the right approach here, since they don't expire. I also
want our stars to be much smarter over time. It's not a static one time
thing. We can customize the look of profiles and stars based on
registration status and maybe even how long they have been a continuous
subscriber, like streak rewards.

Sounds good, I will make the app consult with the server through the REST API as it is simple and reliable _(at least until or if we find a better approach later on)_

**Comment by danieldaquino** [comment]
Created a welcome screen to be displayed when the user successfully purchases the subscription:

https://github.com/damus-io/damus/assets/24692108/638870f8-b582-4a1f-89cf-c619327f9e03

I tried to compose the view and the animation in a way that plays with purple and Damus colors, stars _(since that's the badge symbol they get on their profile)_ and a bit of suspense-building by delaying the rest of the content.

@jb55, @alltheseas, @robagreda please let me know if this looks good, or if you have any suggestions. I will keep working on the rest of the ticket in the meantime. Thanks!

**Comment by alltheseas** [comment]
🤩

**Comment by danieldaquino** [comment]
Just made one other quick improvement to the animation: Added a bit of an animation effect to the background stars for a more immersive feeling

This is the latest:

https://github.com/damus-io/damus/assets/24692108/9a031fcc-d7b1-4de0-bb52-6205f2c4fbc9

**Comment by jb55** [comment]
On Mon, Dec 04, 2023 at 06:18:04PM -0800, Daniel D’Aquino wrote:
>Just made one other quick improvement to the animation: Added a bit of
>an animation effect to the background stars for a more immersive
>feeling
>
>This is the latest:
>
>https://github.com/damus-io/damus/assets/24692108/9a031fcc-d7b1-4de0-bb52-6205f2c4fbc9

that looks awesome, what did you use for this?

**Comment by danieldaquino** [comment]
> that looks awesome, what did you use for this?

Thanks! 

This is pure SwiftUI, using various animation modifiers for each item. 

The star background was made using Pixelmator Pro, and the mesh gradient background was pulled from Rob's Figma designs

**Comment by danieldaquino** [comment]
## Updates on this project for today and Wednesday

1. Improved the infrastructure around the server code and automated testing to speed up development _(Testing manually at this early stage proved to be difficult and unreliable)_
2. Created essential API endpoints for account management (creation, verification, etc)
3. Created an API endpoint for receiving and verifying App Store receipts. 
    - For local testing I have to mock the behaviour
    - I have researched and drafted some logic for the real secure receipt verification with Apple, but I haven't completed or tested it yet
4. Created the logic for checking the user status and displaying the star on the profile
5. Created the logic for collecting and sending over the App Store receipts to the server.
6. Created a Sandbox account for testing, and studied more about the App Store Sandbox environment

## What we have working so far

**Overall, this is the flow that we have working on a local test environment:**
1. User goes to the Purple page _(95% nicely designed, a few quirks)_
2. User chooses a subscription option
3. User confirms the simulated purchase with Apple
4. User is shown the welcome message
5. Device communicates with the API server and registers the user by sending the receipt.
    - **Note:** Receipt verification in this example is bogus, it just accepts any receipt for now
6. User gets a star on their profile, and if they were to call the translation API route, it would allow them to translate. (i.e. they are registered on the server)

## Code changes so far:

- **Damus code:** https://github.com/damus-io/damus/compare/master...danieldaquino:damus:1422-damus-purple-draft-v1
- **Damus API code:** https://github.com/damus-io/api/compare/master...1422-damus-purple-draft-v1

## Next steps

1. Implement npub authentication for all account-related actions
2. Get this flow working with a Sandbox Apple ID account for better and more realistic testing
3. Get it working with a real App Store receipt verification _(I might need @jb55's help for some of the configuration required for this)_
3. Figure out remaining quirks and edge cases before shipping the feature

I have updated the acceptance criteria. Some items that I have added could potentially be delegated to separate tickets.

@jb55, @alltheseas, there are a lot of things going on in this ticket, but hopefully this message provides a good overview of where we are! Please let me know if there are any questions or suggestions regarding any of this.

Thanks!
Daniel

**Comment by danieldaquino** [comment]
Updated the acceptance criteria and delegated various items to separate smaller tickets and more narrowly-scoped tickets to help us better manage this feature: https://github.com/damus-io/damus/milestone/11

@jb55, @alltheseas, please feel free to add/edit/remove stuff from that milestone as needed!

I will wrap up this ticket, send the patch very soon, and start working on other tickets in that milestone (Starting with https://github.com/damus-io/damus/issues/1809)

**Comment by danieldaquino** [comment]
Sent patches!
- **Damus:** https://groups.google.com/a/damus.io/g/patches/c/kNPftAn7yE8
- **Damus API:** https://groups.google.com/a/damus.io/g/patches/c/DdrgniIKavY
